### PR TITLE
Update the CNI's DNAT rule when port is bound to 127.0.0.1

### LINKED ISSUE #4876: Port binding to localhost is not working with nerdctl
When using containerd backend (nerdctl) and publishing ports using the localhost:
```
nerdctl run -d --restart=always --publish 127.0.0.1:8333:80 nginx:stable
```
The application is not accessible from the host. This occurs on both the default network and the new Rancher Desktop Networking. Tracing back to this issue, it seems it was introduced in `1.8.1`.

### PR DESCRIPTION
**Pull request by Nino-K**
Fixes: https://github.com/rancher-sandbox/rancher-desktop/issues/4876

### CODE REVIEW

File: pkg/containerd/events.go [hunk 1]
```diff
@@ -97,6 +112,20 @@ func (e *EventMonitor) MonitorPorts(ctx context.Context) {
 					continue
 				}
 
+				// We only need to perform this check
+				// when the namespaced network is enabled
+				// in the case of default network when the
+				// privileged service is disabled, we will
+				// not be using this event API.
```
Comment (mook-as) [inline pkg/containerd/events.go]: This is a bit of a run-on sentence (a full 5 lines!)
```suggestion
				// We only need to perform this check
				// when the namespaced network is enabled.
				// When using the default network without
				// the privileged service, we will
				// not be using this event API.
```
It's missing _why_ we're not using that event API, though.
Also, it doesn't say anything about default network + privileged service…

File: pkg/containerd/events.go [hunk 2]
```diff
@@ -97,6 +112,20 @@ func (e *EventMonitor) MonitorPorts(ctx context.Context) {
...
+				if !e.enablePrivilegedService {
+					if err = execIptablesRules(
+						ports, startTask.ContainerID,
+						envelope.Namespace,
+						strconv.Itoa(int(startTask.Pid))); err != nil {
+						log.Errorf("failed running iptable rules to update DNAT rule in CNI-HOSTPORT-DNAT chain: %v", err)
```
Comment (mook-as) [inline pkg/containerd/events.go:122]: This indenting is unreadble.

Please split this to two lines (`err = execIptablesRules(…); if err != nil {`) due to line length reasons.

(I was going to say "move the brace to start of line to make it more readable", but splitting into two lines is clearer.)

Please also file an issue for a refactor (to make the contents of `case "/tasks/start":` a new helper method). I assume that's doable, at least; if it looks unreasonable we can skip that (though it's not great).

File: pkg/containerd/events.go [hunk 3]
```diff
@@ -206,6 +235,91 @@ func (e *EventMonitor) Close() error {
 	return finalErr
 }
 
+// When namespaced network is enable, if portbinding is bound to 127.0.0.1 we update the DNAT
```
Comment (mook-as) [inline pkg/containerd/events.go]:
```suggestion
// When namespaced network is enabled, if portbinding is bound to 127.0.0.1 we update the DNAT
```

File: pkg/containerd/events.go [hunk 4]
```diff
@@ -206,6 +235,91 @@ func (e *EventMonitor) Close() error {
...
+// in main CNI DNAT chain (CNI-HOSTPORT-DNAT). The existing DNAT rule are as follows:
```
Comment (mook-as) [inline pkg/containerd/events.go]: Please rephrase the first sentence; it's missing a _why_ we do things.

File: pkg/containerd/events.go [hunk 5]
```diff
@@ -206,6 +235,91 @@ func (e *EventMonitor) Close() error {
...
+// DNAT       tcp  --  anywhere             localhost            tcp dpt:9119 to:10.4.0.22:80.
+// While, this rule only allows the traffic to get routed to the localhost only we update
+// the rule to allow the traffic to any destination IP address.
```
Comment (mook-as) [inline pkg/containerd/events.go]:
```suggestion
// While this rule only allows the traffic to get routed to the localhost only we update
// the rule to allow the traffic to any destination IP address.
```
Actually, maybe rephrase it more…
```suggestion
// We modify the rule to instead capture all traffic to that port regardless of the
// destination address:
```

This seems to lead to issues that I'm not sure how to resolve; imagine a user has two services, listening on `127.0.0.1:9999` and `0.0.0.0:9999`. This change means we can only have _one_ service instead of two…

File: pkg/containerd/events.go [hunk 6]
```diff
@@ -206,6 +235,91 @@ func (e *EventMonitor) Close() error {
...
+// DNAT       tcp  --  anywhere             anywhere             tcp dpt:9119 to:10.4.0.22:80.
+func execIptablesRules(portMappings nat.PortMap, containerID, namespace, pid string) error {
+	var errs []error
+
+	for portProto, portBindings := range portMappings {
+		for _, portBinding := range portBindings {
+			if portBinding.HostIP == "127.0.0.1" {
+				if err := createLoopbackIPtablesRules(
```
Comment (mook-as) [inline pkg/containerd/events.go]: As above, please don't use `if err = foo(); err != nil {` if the the statement spans lines.

File: pkg/containerd/events.go [hunk 7]
```diff
@@ -206,6 +235,91 @@ func (e *EventMonitor) Close() error {
 	return finalErr
 }
 
+// When namespaced network is enable, if portbinding is bound to 127.0.0.1 we update the DNAT
+// in main CNI DNAT chain (CNI-HOSTPORT-DNAT). The existing DNAT rule are as follows:
+// DNAT       tcp  --  anywhere             localhost            tcp dpt:9119 to:10.4.0.22:80.
+// While, this rule only allows the traffic to get routed to the localhost only we update
+// the rule to allow the traffic to any destination IP address.
+// DNAT       tcp  --  anywhere             anywhere             tcp dpt:9119 to:10.4.0.22:80.
+func execIptablesRules(portMappings nat.PortMap, containerID, namespace, pid string) error {
```
Comment (mook-as) [inline pkg/containerd/events.go:271]: While the comment describes what this method is supposed to do, it doesn't relate to the inputs. This comment is probably more appropriate for `createLoopbackIPtablesRules` instead? That does mean `execIptablesRules` still needs a better comment, though.

File: pkg/containerd/events.go [hunk 8]
```diff
@@ -287,6 +401,45 @@ func createPortMappingFromString(portMapping string) (nat.PortMap, error) {
...
+	if len(matches) < segments {
+		return "", ErrIPAddressNotFound
+	}
+
+	return matches[1], nil
+}
+
+// ported from:
+// https://github.com/containernetworking/plugins/blob/2b097c5a62dacfd2f9ea4268faa4fc04bd5343f5/pkg/utils/utils.go#L39
+// however module requires Go 1.20
```
Comment (mook-as) [inline pkg/containerd/events.go:419]: Is there a reason we don't just bump up the requirements?

File: pkg/containerd/events.go [hunk 9]
```diff
@@ -287,6 +401,45 @@ func createPortMappingFromString(portMapping string) (nat.PortMap, error) {
...
+// MustFormatChainNameWithPrefix generates a chain name similar
+// to FormatChainName, but adds a custom prefix between
+// chainPrefix and unique identifier. Ensures that the
+// generated chain name is exactly maxChainLength chars in length.
+// Panics if the given prefix is too long.
+func mustFormatChainNameWithPrefix(name string, id string, prefix string) string {
+	return mustFormatHashWithPrefix(maxChainLength, chainPrefix+prefix, name+id)
```
Comment (mook-as) [inline pkg/containerd/events.go]: If we're copying the code (i.e. we can't just upgrade golang), we should skip the extra layer and just hard-code the prefix (since that's always fixed for us).

File: pkg/containerd/events.go [hunk 10]
```diff
@@ -287,6 +401,45 @@ func createPortMappingFromString(portMapping string) (nat.PortMap, error) {
 	return portMap, nil
 }
 
+func extractIPAddress(line string) (string, error) {
+	// Regular expression pattern to match the IP address
+	ipPattern := `inet\s+(\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3})\/\d{1,2}`
```
Comment (mook-as) [inline pkg/containerd/events.go]:
```suggestion
	ipPattern := `\binet\s+(\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3})\/\d{1,2}`
```

File: pkg/containerd/events.go [hunk 11]
```diff
@@ -287,6 +401,45 @@ func createPortMappingFromString(portMapping string) (nat.PortMap, error) {
 	return portMap, nil
 }
 
+func extractIPAddress(line string) (string, error) {
```
Comment (mook-as) [inline pkg/containerd/events.go]: This could use a doc comment on what this does. Or maybe move running `nsenter ip` into this function.

File: pkg/containerd/events.go [hunk 12]
```diff
@@ -206,6 +235,91 @@ func (e *EventMonitor) Close() error {
...
+	if len(errs) != 0 {
+		return fmt.Errorf("%w: %+v", ErrExecIptablesRule, errs)
+	}
+
+	return nil
+}
+
+func createLoopbackIPtablesRules(containerID, namespace, pid, port, destinationPort string) error {
+	// read the container's cni config to extract the network name
+	nsenterNetworkConfCmd := exec.Command("nsenter", "-t", pid, "-n", "cat", "/etc/cni/net.d/nerdctl-bridge.conflist")
```
Comment (mook-as) [inline pkg/containerd/events.go:305]: If we're hard-coding `nerdctl-bridge.conflist` (i.e. this is irrespective of which network the container is actually in), couldn't we have hard-coded `bridge` here too? Or am I missing something here (likely)?

**Review by mook-as — CHANGES_REQUESTED** [review]

File: pkg/containerd/events.go (hunk 5 above)
Comment (Nino-K) [reply pkg/containerd/events.go]: Users won't be able to re-use `9999` port twice the host-switch proxy will not allow them and will get an error.

File: pkg/containerd/events.go (hunk 7 above)
Comment (Nino-K) [reply pkg/containerd/events.go:271]: You are right, I ended up moving things around and left the comment behind.

File: pkg/containerd/events.go (hunk 8 above)
Comment (Nino-K) [reply pkg/containerd/events.go:419]: I did bump initially, however, that requires our environment to also get updated. Given the time crunch, I thought it's best to leave it for 1.10.

File: pkg/containerd/events.go (hunk 12 above)
Comment (Nino-K) [reply pkg/containerd/events.go:305]: Created an issue: https://github.com/rancher-sandbox/rancher-desktop-agent/issues/47

File: pkg/containerd/events.go [hunk 13]
```diff
@@ -253,6 +267,98 @@ func (e *EventMonitor) updateListener(
...
+
+	return nil
+}
+
+// When namespaced network is enabled, if the port binding is bound to 127.0.0.1, we add an additional
+// DNAT rule in the main CNI DNAT chain (CNI-HOSTPORT-DNAT) after the existing rule (using --append).
+// This is necessary because the initial CNI rule created by containerd only allows the traffic
+// to be routed to localhost. Therefore, we add an additional rule to allow traffic to any
+// destination IP address which allows the service to be discoverable through namespaced network's
+// subnet, which essentially causes the service to listen on eth0 instead; this is required as the
```
Comment (mook-as) [inline pkg/containerd/events.go:297]: I don't think subnets are really involved here.

File: pkg/containerd/events.go [hunk 14]
```diff
@@ -253,6 +267,98 @@ func (e *EventMonitor) updateListener(
...
+	}
+
+	log.Debugf("found the ip address: %s for containerID: %s", eth0IP, containerID)
+
+	// create the corresponding chanin name, e.g CNI-DN-xxxxxx
+	// (where xxxxxx is a function of the ContainerID and network name)
+	// https://www.cni.dev/plugins/current/meta/portmap/#dnat
+	// https://github.com/containernetworking/plugins/blob/2b097c5a62dacfd2f9ea4268faa4fc04bd5343f5/pkg/utils/utils.go#L39
+	cID := fmt.Sprintf("%s-%s", namespace, containerID)
+	chainName := mustFormatHashWithPrefix(maxChainLength, chainPrefix+"DN-", networkConfig.Name+cID)
```
Comment (mook-as) [inline pkg/containerd/events.go:331]: It's odd that `chainPrefix` is a constant that's only used here, concatenated with a second string. But we can fix that later.

**Review by mook-as — APPROVED** [review]
Eh close enough
